File: dns/entry.rb
require 'json'
require 'yaml'
require 'aws-sdk'

def stack_output(value)
  _, stack, key = value.split(':')
  outputs = Aws::CloudFormation::Stack.new(name: stack).outputs
  outputs.find { |o| o.output_key == key }.output_value
end

pillar = {}
pillar['key'] = ENV['CLOUDFLARE_API_KEY']
pillar['email'] = ENV['CLOUDFLARE_EMAIL']

zones = YAML.load_file('./zones.yml')
zones.each do |_, info|
  info.fetch('records', []).each do |r|
    if r.fetch('content', '').start_with? 'stack_output:'
      r['content'] = stack_output(r['content'])
    end
  end
end
pillar['zones'] = zones

system('salt-call --local saltutil.sync_states')
system("salt-call --local state.sls dns --state-output=mixed pillar='#{pillar.to_json}'")

File: dns/entry.py
import os
import json

import yaml
import boto3


def stack_output(content):
    cloudformation = boto3.resource('cloudformation')
    _, stack, key = content.split(':')
    outputs = cloudformation.Stack(stack).outputs

    for output in outputs:
        if output['OutputKey'] == key:
            return output['OutputValue']

    raise Exception('Could not find output "{}" in stack "{}"'.format(key, stack))


def main():
    zones = yaml.load(open('./zones.yml'))

    import ipdb; ipdb.set_trace()
    for zone, info in zones.iteritems():
        for record in info.get('records', []):
            if record.get('content', '').startswith('stack_output:'):
                record['content'] = stack_output(record['content'])

    pillar = json.dumps({
        'key': os.environ['CLOUDFLARE_API_KEY'],
        'email': os.environ['CLOUDFLARE_EMAIL'],
        'zones': zones
    })

    os.system('salt-call --local saltutil.sync_states')
    os.system("salt-call --local state.sls dns --state-output=mixed pillar='{}'".format(pillar))


if __name__ == '__main__':
    main()
